src/Language/PureScript/TypeChecker/Roles.hs

```diff
@@ -80,72 +84,87 @@ lookupRoles
 lookupRoles env tyName =
   fromMaybe [] $ M.lookup tyName (types env) >>= typeKindRoles . snd
 
--- | This function does the following:
---
--- * Infers roles for the given data type declaration
---
--- * Compares the inferred roles to the explicitly declared roles (if any) and
---   ensures that the explicitly declared roles are not more permissive than
---   the inferred ones
+-- |
+-- Compares the inferred roles to the explicitly declared roles and ensures
+-- that the explicitly declared roles are not more permissive than the
+-- inferred ones.
 --
 checkRoles
   :: forall m
    . (MonadError MultipleErrors m)
-  => Environment
+  => [(Text, Maybe SourceType, Role)]
+    -- ^ type parameters for the data type whose roles we are checking
+  -> [Role]
+    -- ^ roles declared for the data type
+  -> m ()
+checkRoles tyArgs declaredRoles = do
+  let k (var, _, inf) dec =
+        when (inf < dec) . throwError . errorMessage $ RoleMismatch var inf dec
```

Comment:
I think this error message could be improved to indicate which type parameter's role was wrong in a declaration with 2+ role annotations. For example, looking at this error message
- [CoercibleRoleMismatch1.purs](https://github.com/purescript/purescript/blob/146b795fb9084994bd12b971dd559b2a1e0911f4/tests/purs/failing/CoercibleRoleMismatch1.purs)
- [CoercibleRoleMismatch1.out](https://github.com/purescript/purescript/pull/4157/files?w=0#diff-ba8180994d237116bc772b4e39a6e31fc4c9f5ca66e3ce14602d4614acd23c6cR3)

... the source spans indicates the entire role declaration. They do not indicate which role declaration has the issue.

If I have the following example,
```purescript
data Foo a b c d = Foo a b c

type role Foo representational phantom representational phantom
```
and `b` is the one with the bad role, the error message wouldn't tell me which of the four values above is wrong.

Reply:
On second thought, the issue above is a separate issue. This PR AFAICT accomplishes its goal of checking for arity. I'll retract this.